### ISSUE TITLE
TimeZoneRegistry.getTimeZone always returns null

### ISSUE THREAD

**Issue by DanielNovo** [ical4j/ical4j#219]
With version 2.1.5

```
TimeZoneRegistry registry = TimeZoneRegistryFactory.getInstance().createRegistry();
TimeZone timezone = registry.getTimeZone("Portugal");
```

timezone is always null.

After debugging source code, TimeZoneLoader silently fails in method loadVTimeZone, line 99:

`final Calendar calendar = builder.build(resource.openStream());`

this fails with IOException: Pushback buffer overflow.

Am i missing something? Can you please check this?

Thanks in advance.

**Comment by dbularzik** [comment]
I ran into the same issue when I configured ical4j to use HCalendarParserFactory, as suggested by the wiki in the Configuration section. This is because the timezone definitions included in the ical4j release (and ical4j-zoneinfo-outlook) are NOT in hCalendar (XHTML) format, and the code uses the same parser factory regardless of whether it's reading an actual calendar or just the .ics files that define timezones. 

(Note it took some work to figure out that this was the case--since the UnfoldingReader throws pushback buffer overflow exceptions when the patterns it's trying to unfold occur somewhere other than the end of the read buffer...which is what happens when the code expects XHTML but gets ics text. It would have been easier to debug if the pushback buffer was set to something larger than 3.)

To work around this issue, simply set the following property in your ical4j properties file:

net.fortuna.ical4j.parser=net.fortuna.ical4j.data.DefaultCalendarParserFactory

...though realize by doing so you lose the ability to read hCalendar format. 

Alternately, you could replace the timezone definitions (bundled in the jarfiles that are in Maven Central, so you'd have to build locally) with hCalendar equivalents.

The true fix for this would be to have separate parser properties for timezones and calendars.

**Comment by benfortuna** [comment]
Hi Daniel,

Timezones aren't defined by country, but rather follow the Olson DB format of well-known (often capital) cities. So for Portugese time you might use "Europe/Lisbon".

Search for Olson DB for more info or have a look at http://tzurl.org

I think the HCalendar issue is something entirely different.

**Comment by dbularzik** [comment]
Right...my issue may have been slightly different than the OP. But if I used, say, "America/New_York" (a valid timezone from the Olson DB), AND I have my parser set to hCalendar, I would see the same error as the OP. Debugging revealed that the root cause was that the timezone files it was reading as a resource from the local jar was in ICS format, not XHTML. The UnfoldingReader assumes that the characters that it will have to push back will only happen at the end of the line that's read in; it therefore has a small pushback buffer, large enough to only handle the character sequence it's looking for. However, when it's reading an ICS file (which is essentially text), it can encounter that sequence in the middle of the 64-byte chunk that it reads, and therefore can easily exceed the buffer size when it tries to push back the rest of the chunk. 

However, if you change the pushback buffer size in UnfoldingReader to something large enough to handle the entire chunk, you then get a Xerxes parsing exception, due to the fact that the ICS file is not XML. (More specifically, it considers the text in the ICS file to be an invalid prolog to XML.)

Switching the parser to the DefaultCalendarParser fixes this, at the cost of losing the ability to read hCalendar files. Not such a big deal to me; this library is still incredibly useful without that feature. But it seems that right now you have a choice to either be able to read the timezone ICS files from resources, OR you can support hCalendar, not both.

**Comment by leventepal** [comment]
I'm having the same issue (Android):

```
TimeZoneRegistryFactory tzrf = TimeZoneRegistryFactory.getInstance();
TimeZoneRegistry tzr = tzrf.createRegistry();

net.fortuna.ical4j.model.TimeZone timeZone = tzr.getTimeZone("Europe/Bucharest");
```

It's returning null.

Android configuration:
```
implementation 'org.mnode.ical4j:ical4j:2.2.0'
implementation group: 'org.jsr107.ri', name: 'cache-ri-impl', version: '1.1.0'
implementation group: 'javax.cache', name: 'cache-api', version: '1.1.0'
```

Proguard:
```
-dontwarn org.apache.commons.logging.**
-keep class org.apache.commons.logging.** { *; }
-dontwarn sun.misc.Perf
-keep class sun.misc.Perf { *; }
-dontnote com.google.vending.**
-keep class com.google.vending.** { *; }
-dontnote com.android.vending.licensing.**
-keep class com.android.vending.licensing.** { *; }
-dontwarn net.fortuna.ical4j.model.**
-keep class net.fortuna.ical4j.model.** { *; }
-dontwarn org.jsr107.**
-keep class org.jsr107.** { *; }
-keep class javax.cache.** { *; }
-dontwarn java.cache.**
-keep class org.jsr107.ri.** { *; }
-dontwarn org.jsr107.ri.**

-dontwarn javax.cache.management.**
-keep class javax.cache.management.** { *; }
-dontwarn javax.enterprise.inject.spi.**
-keep class javax.enterprise.inject.spi.** { *; }

-dontwarn aQute.bnd.**
-dontwarn groovy.**
-dontwarn org.codehaus.**
-dontwarn groovyjarjarasm.**
-dontwarn groovyjarjarantlr.**
-dontwarn org.slf4j.**
```

**Comment by Bartosz-D3V** [comment]
I have exactly the same issue - running the following code results in NullPointerException:
```java
final TimeZoneRegistry registry = TimeZoneRegistryFactory.getInstance().createRegistry();
final TimeZone timezone = registry.getTimeZone("Europe/London");
```
Variable timezone is always null.

**Comment by rfc2822** [comment]
See #229 for another stack trace

**Comment by noloman** [comment]
same here, but trying to retrieve "Europe/Amsterdam"

**Comment by DanielNovo** [comment]
Thanks dbularzik.

Changing property net.fortuna.ical4j.parser to net.fortuna.ical4j.data.DefaultCalendarParserFactory fixed my issue.

**Comment by schube** [comment]
Thank you DanielNovo !

>Changing property net.fortuna.ical4j.parser to net.fortuna.ical4j.data.DefaultCalendarParserFactory fixed my issue.

This finally fixed the problem for me, too!

**Comment by decrecondiz** [comment]
Im currently having the same issue and have tried the below to no avail

> Thank you DanielNovo !
> 
> > Changing property net.fortuna.ical4j.parser to net.fortuna.ical4j.data.DefaultCalendarParserFactory fixed my issue.
> 
> This finally fixed the problem for me, too!

This has been tried with multiple timezones ("Etc/GMT+0","Europe/London","America/Mexico_City", etc) all returning the same error.

Code sample:
![returns null](https://user-images.githubusercontent.com/17993127/101282215-acaf7200-37cb-11eb-9243-205fcf517e6d.jpg)

Stack trace:
![stack trace](https://user-images.githubusercontent.com/17993127/101282260-fc8e3900-37cb-11eb-8423-ab5b4c0e342e.jpg)

**Comment by findweek-xgi** [comment]
On Linux, a '/' should be added to  "zoneinfo/" of String DEFAULT_RESOURCE_PREFIX in TimeZoneRegistryImpl to make it work. I think it should be added to have this code work on all platforms

**Comment by schteff** [comment]
DefaultCalendarParserFactory seems to have been removed between versions 3.0.4 and 3.0.5. Is there any recommended fix for newer versions?

**Comment by benfortuna** [comment]
To get a new parser instance you can use CalendarParserFactory directly:

```java
CalendarParser parser = CalendarParserFactory.getInstance();
```

This will return an instance of the configured custom parser (via proerty `net.fortuna.ical4j.parser`), or `CalendarParserImpl` by default.

**Comment by schteff** [comment]
Thanks! It seems like the null problem with registry.getTimeZone("") no longer happens in the latest version. I think you can close this issue. I just removed the net.fortuna.ical4j.parser property and get no errors.
